qiskit_experiments/database_service/utils.py

```diff
@@ -276,3 +279,250 @@ def append(self, value):
+            default_value: Default value to fill added columns.
+        """
+        # Order sensitive
+        new_columns = [c for c in new_columns if c not in self.get_columns()]
+        self._extra.extend(new_columns)
+
+        # Update current table
+        with self._lock:
+            for new_column in new_columns:
+                self._container.insert(len(self._container.columns), new_column, default_value)
+        self._columns.extend(new_columns)
+
+    def clear(self):
+        """Remove all elements from this container."""
+        with self._lock:
+            self._container = self._init_container()
+            self._columns = self._default_columns()
+            self._extra = []
+
+    def container(
+        self,
+        collapse_extra: bool = True,
+    ) -> pd.DataFrame:
+        """Return bare pandas dataframe.
+
+        Args:
+            collapse_extra: Set True to show only default columns.
+
+        Returns:
+            Bare pandas dataframe. This object is no longer thread safe.
+        """
+        with self._lock:
+            container = self._container
+
+        if collapse_extra:
+            return container[self._default_columns()]
+        return container
+
+    def add_entry(self, **kwargs):
+        """Add new entry to the dataframe.
+
+        Args:
+            kwargs: Description of new entry to register.
+        """
+        columns = self.get_columns()
+        missing = kwargs.keys() - set(columns)
+        if missing:
+            self.add_columns(*sorted(missing))
+
+        template = dict.fromkeys(self.get_columns())
+        template.update(kwargs)
+
+        if not template["result_id"]:
+            template["result_id"] = uuid.uuid4().hex
+        name = self._unique_table_index(template["result_id"])
+        with self._lock:
+            self._container.loc[name] = list(template.values())
+
+    def _unique_table_index(self, index_name: str):
+        """Generate unique index name with 8 characters."""
+        if not isinstance(index_name, str):
+            index_name = str(index_name)
+        truncated = index_name[:8]
+        with self.lock:
+            while truncated in self._container.index:
+                truncated = uuid.uuid4().hex[:8]
+        return truncated
+
+    def _repr_html_(self) -> Union[str, None]:
+        """Return HTML representation of this dataframe."""
+        with self._lock:
+            # Remove underscored columns.
+            return self._container._repr_html_()
+
+    def __getattr__(self, item):
+        lock = object.__getattribute__(self, "_lock")
+
+        with lock:
+            # Lock when access to container's member.
+            container = object.__getattribute__(self, "_container")
+            if hasattr(container, item):
+                return getattr(container, item)
+        raise AttributeError(f"'ThreadSafeDataFrame' object has no attribute '{item}'")
+
+    def __json_encode__(self) -> Dict[str, Any]:
+        return {
```

Comment:
The entire dataframe is JSON serializable with the experiment encoder/decoder. So you can call save API for the batch experiment results in one piece. This drastically improve the data saving performance once experiment service updates its API.